File: test_samples/comprehensive/Complex.php
<?php
// Comprehensive PHP test file with complex structures

namespace Authentication;

/**
 * Authenticator interface
 */
interface Authenticator {
    public function authenticate(string $username, string $password): bool;
    public function logout(): void;
}

/**
 * Loggable trait
 */
trait Loggable {
    protected function log(string $message): void {
        echo "[LOG] $message\n";
    }
}

/**
 * Base User class
 */
abstract class User {
    protected string $username;
    protected int $id;
    
    public function __construct(string $username, int $id) {
        $this->username = $username;
        $this->id = $id;
    }
    
    abstract public function getRole(): string;
    
    public function getUsername(): string {
        return $this->username;
    }
}

/**
 * AdminUser class with trait and interface
 */
class AdminUser extends User implements Authenticator {
    use Loggable;
    
    private array $permissions = [];
    
    public function __construct(string $username, int $id) {
        parent::__construct($username, $id);
        $this->log("Admin user created: $username");
    }
    
    public function getRole(): string {
        return 'ADMIN';
    }
    
    public function authenticate(string $username, string $password): bool {
        if (empty($username) || empty($password)) {
            return false;
        }
        return $this->username === $username && $this->validatePassword($password);
    }
    
    public function logout(): void {
        $this->log("Admin logged out: {$this->username}");
    }
    
    public function addPermission(string $permission): void {
        $this->permissions[] = $permission;
        $this->log("Permission added: $permission");
    }
    
    private function validatePassword(string $password): bool {
        return strlen($password) >= 8;
    }
}

/**
 * Generic SessionManager class
 */
class SessionManager {
    private static ?SessionManager $instance = null;
    private array $sessions = [];
    
    private function __construct() {}
    
    public static function getInstance(): SessionManager {
        if (self::$instance === null) {
            self::$instance = new SessionManager();
        }
        return self::$instance;
    }
    
    public function createSession(string $sessionId, User $user): void {
        $this->sessions[$sessionId] = $user;
    }
    
    public function getSession(string $sessionId): ?User {
        return $this->sessions[$sessionId] ?? null;
    }
    
    public function removeSession(string $sessionId): void {
        unset($this->sessions[$sessionId]);
    }
}

/**
 * StatusCode enum (PHP 8.1+)
 */
enum StatusCode: int {
    case SUCCESS = 200;
    case UNAUTHORIZED = 401;
    case FORBIDDEN = 403;
    case NOT_FOUND = 404;
    
    public function getMessage(): string {
        return match($this) {
            self::SUCCESS => 'OK',
            self::UNAUTHORIZED => 'Unauthorized',
            self::FORBIDDEN => 'Forbidden',
            self::NOT_FOUND => 'Not Found',
        };
    }
}

/**
 * Complex function with type hints
 */
function authenticateAndCreateSession(
    Authenticator $user,
    string $username,
    string $password,
    SessionManager $manager
): array {
    if ($user->authenticate($username, $password)) {
        $sessionId = uniqid('session_', true);
        $manager->createSession($sessionId, $user);
        return ['success' => true, 'sessionId' => $sessionId];
    }
    return ['success' => false, 'error' => StatusCode::UNAUTHORIZED->getMessage()];
}

File: test_samples/comprehensive/complex.py
# Comprehensive Python test file with complex structures
from abc import ABC, abstractmethod
from dataclasses import dataclass, field
from typing import Dict, List, Optional, Generic, TypeVar
from threading import Lock
from enum import Enum

class Authenticator(ABC):
    """Abstract authenticator interface"""
    
    @abstractmethod
    def authenticate(self, username: str, password: str) -> bool:
        pass
    
    @abstractmethod
    def logout(self) -> None:
        pass


@dataclass
class User(ABC):
    """Base user class with dataclass"""
    username: str
    id: int
    _lock: Lock = field(default_factory=Lock, init=False, repr=False)
    
    @abstractmethod
    def get_role(self) -> str:
        pass


class AdminUser(User, Authenticator):
    """AdminUser with multiple inheritance"""
    
    def __init__(self, username: str, id: int):
        super().__init__(username=username, id=id)
        self._permissions: List[str] = []
    
    def get_role(self) -> str:
        return "ADMIN"
    
    def authenticate(self, username: str, password: str) -> bool:
        if not username or not password:
            return False
        return self.username == username and self._validate_password(password)
    
    def logout(self) -> None:
        print("Admin logged out")
    
    def add_permission(self, permission: str) -> None:
        with self._lock:
            self._permissions.append(permission)
    
    def get_permissions(self) -> List[str]:
        with self._lock:
            return self._permissions.copy()
    
    def _validate_password(self, password: str) -> bool:
        return len(password) >= 8


T = TypeVar('T', bound=User)

class SessionManager(Generic[T]):
    """Generic session manager using singleton pattern"""
    
    _instance: Optional['SessionManager'] = None
    _lock: Lock = Lock()
    
    def __new__(cls):
        if cls._instance is None:
            with cls._lock:
                if cls._instance is None:
                    cls._instance = super().__new__(cls)
                    cls._instance._sessions = {}
        return cls._instance
    
    def create_session(self, session_id: str, user: T) -> None:
        with self._lock:
            self._sessions[session_id] = user
    
    def get_session(self, session_id: str) -> Optional[T]:
        with self._lock:
            return self._sessions.get(session_id)
    
    def remove_session(self, session_id: str) -> None:
        with self._lock:
            self._sessions.pop(session_id, None)


class StatusCode(Enum):
    """Status code enum"""
    SUCCESS = 200
    UNAUTHORIZED = 401
    FORBIDDEN = 403
    NOT_FOUND = 404
    
    @property
    def message(self) -> str:
        messages = {
            StatusCode.SUCCESS: "OK",
            StatusCode.UNAUTHORIZED: "Unauthorized",
            StatusCode.FORBIDDEN: "Forbidden",
            StatusCode.NOT_FOUND: "Not Found",
        }
        return messages.get(self, "Unknown")


def authenticate_and_create_session(
    user: Authenticator,
    username: str,
    password: str,
    manager: SessionManager
) -> Dict[str, any]:
    """Complex function with type hints"""
    if user.authenticate(username, password):
        import uuid
        session_id = f"session_{uuid.uuid4()}"
        manager.create_session(session_id, user)
        return {"success": True, "session_id": session_id}
    return {"success": False, "error": StatusCode.UNAUTHORIZED.message}


# Context manager
class AuthenticationContext:
    """Context manager for authentication"""
    
    def __init__(self, user: Authenticator):
        self.user = user
    
    def __enter__(self):
        return self.user
    
    def __exit__(self, exc_type, exc_val, exc_tb):
        self.user.logout()
        return False


# Decorator
def require_authentication(func):
    """Decorator for authentication requirement"""
    def wrapper(self, *args, **kwargs):
        if not hasattr(self, 'authenticated') or not self.authenticated:
            raise PermissionError("Authentication required")
        return func(self, *args, **kwargs)
    return wrapper
